source/agora/common/crypto/Schnorr.d

```diff
@@ -64,6 +64,481 @@ nothrow @nogc unittest
+
+        public void deserialize (scope DeserializeDg dg) @safe
+        {
+            deserializePart(this.validators, dg);
+            deserializePart(this.signature, dg);
+
+            size_t length;
+            deserializePart(length, dg);
+
+            foreach (_; 0 .. length)
+            {
+                ushort idx;
+                deserializePart(idx, dg);
+
+                Hash preimage;
+                deserializePart(preimage, dg);
+                this.preimages[idx] = preimage;
+            }
+
+            deserializePart(this.data, dg);
+        }
+    }
+
+    // 8 validators
+    Pair[] pair_keys;
+    8.iota.each!(_ => pair_keys ~= Pair.random());
+
+    Pair[] genNodesRands (size_t length)
+    {
+        Pair[] nodes_rands;
+        nodes_rands.length = 8;
+
+        foreach (ref pair; nodes_rands)
+            pair = Pair.random();
+
+        return nodes_rands;
+    }
+
+    // these are the initial R's that were precommited in the enrollment
+    Pair[] nodes_rands = genNodesRands(8);
+
+    // the calculated R's for each node for the previous block (always replaced,
+    // but might keep a history of these in case blocks get rolled-back)
+    Point[] prev_pubs;
+    prev_pubs.length = 8;
+
+    // update expected R's based on the previously finalized block
+    void updateExpectedRandoms (Block block)
+    {
+        // update expected previous R's for each node
+        foreach (idx, preimage; block.preimages)
+            prev_pubs[idx] = prev_pubs[idx] + Scalar(preimage).toPoint();
+    }
+
+    // they're sorted alphabetically (for the bitmask indices)
+    Point[] pub_keys = pair_keys.map!(pair => pair.V).array;
+    pub_keys.sort();
+
+    // generates consecutive rounds of hashes for the initial scalar
+    // (preimages in the resulting array should be revealed from the back)
+    Hash[] genPreimages (Scalar s)
+    {
+        Hash[] result;
+        result ~= s.hashFull();
+
+        foreach (_; 0 .. 10)
+            result ~= result[$ - 1].hashFull();
+
+        return result;
+    }
+
+    ushort getIndex (Point key)
+    {
+        auto res = pub_keys.countUntil(key);
+        assert(res != -1);
+        assert(res < ushort.max);
+        return cast(ushort)res;
+    }
+
+    // sanity check
+    assert(getIndex(pub_keys[0]) == 0);
+    assert(getIndex(pub_keys[1]) == 1);
+
+    ///
+    void revealPreimage (ref Block block, Pair key_pair, Hash preimage)
+    {
+        auto signer_index = getIndex(key_pair.V);
+        block.preimages[signer_index] = preimage;
+    }
+
+    ///
+    void signBlock (ref Block block, Pair key_pair, Pair rand_pair)
+    {
+        // the challenge (the Block hash, which includes the preimages)
+        auto msg = block.hashFull();
+
+        auto sig = signChallenge(key_pair.v, rand_pair.V, rand_pair.v, msg);
+
+        block.signature.R = combine(block.signature.R, sig.R);
+        block.signature.s = combine(block.signature.s, sig.s);
```

Comment:
Why not use `block.signature.R + sig.R` ?

Reply:
The problem I had was with the default initializer of Point:

```D
Point p1 = Scalar.random().toPoint;
Point p2 = Scalar.random().toPoint;

struct Block
{
    Point p;
}

Block block;
block.p = block.p + p1;
block.p = block.p + p2;

writefln("p1: %s", p1);
writefln("p2: %s", p2);
writefln("p1 + p2: %s", p1 + p2);
writefln("block.p: %s", block.p);
assert(block.p == p1 + p2);  // fails
```

```
p1: 0x9a59d8f2d8123d144bf646363f441e341352cb34210d09a90c0cab46e0bab355
p2: 0x4a65a474377deb3f200b30a69c5f9789a70f1bef1b7d693bd88f42d645baee2d
p1 + p2: 0xc764ceb8e99503bda74a56fc7806d2154f30bb89034f7e3ea6245f078d5bf668
block.p: 0xdc1a8018205117574df5011b3cc0755313122621c7b1a6d16fa00f2a0267e69a
```

Reply:
Looks like we should disable default construction for `Point`

Reply:
I guess so. The starting point is G, so it's not zero, haha.

Reply:
Actually I think it's G, maybe it's just whatever the binary representation makes it to be..